compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -4438,11 +4438,16 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
         // if the original tree was a lambda. This does not work always either since
         // sometimes we do not have the original anymore and use the transformed tree instead.
         // But taken together, the two criteria are quite accurate.
-        missingArgs(tree, tree.tpe.widen)
+        if isPure then
+          missingArgs(tree, tree.tpe.widen)
+        else
+          report.warning(UnusedNonUnitValue(tree.tpe),original.srcPos)
```

Comment:
A UnusedNonIUnitValue warning would have to be issued by the linter. It's a conscious design decision not to mix typing and linting one one phase. Typing is mandatory., linting is optional. Typing is precise according to spec, whereas linting is ad-hoc and not specified by the language.

Reply:
FWIW, this is a challenge in scala 2, which is more constrained. Typer has information you want to exploit which may not be available later, but you don't want to bloat or break typer; scala 2 resorts to attachments. In this case, I don't care about purity, `-Wnonunit-statement` should always warn. Maybe it won't issue a duplicate warning if the positions are the same.

Reply:
> this is a challenge in scala 2, ... Typer has information you want to exploit which may not be available later

I wonder we can take advantage of Tasty information for linting.  I guess the difficulty would be some linting might depend on compiler state that would not be available from Tasty.